tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,744 @@
+//! # Implementation Details
+//!
+//! The semaphore is implemented using an intrusive linked list of waiters. An
+//! atomic counter tracks the number of available permits. If the semaphore does
+//! not contain the required number of permits, the task attempting to acquire
+//! permits places its waker at the end of a queue. When new permits are made
+//! available (such as by releasing an initial acquisition), they are assigned
+//! to the task at the front of the queue, waking that task if its requested
+//! number of permits is met. 
+//!
+//! Because waiters are enqueued at the back of the linked list and dequeued
+//! from the front, the semaphore is fair. Tasks trying to acquire large numbers
+//! of permits at a time will always be woken eventually, even if many other
+//! tasks are acquiring smaller numbers of permits. This means that in a
+//! use-case like tokio's read-write lock, writers will not be starved by
+//! readers.
+//!
+//! The linked list is guarded by a mutex, which must be acquired before
+//! enqueueing or dequeueing a task. However, some operations are often
+//! wait-free.
+use crate::loom::cell::CausalCell;
+use crate::loom::sync::{atomic::AtomicUsize, Mutex, MutexGuard};
+use crate::util::linked_list::{self, LinkedList};
+
+use std::{cmp, fmt};
+use std::future::Future;
+use std::marker::PhantomPinned;
+use std::pin::Pin;
+use std::ptr::NonNull;
+use std::sync::atomic::Ordering::*;
+use std::task::{Context, Poll, Waker};
+use std::task::Poll::*;
+
+/// An asynchronous counting semaphore which permits waiting on multiple permits at once.
+pub(crate) struct Semaphore {
+    waiters: Mutex<Waitlist>,
+    /// The current number of available permits in the semaphore.
+    permits: AtomicUsize,
+
+    /// Permits in the process of being released.
+    ///
+    /// When releasing permits, if the lock on the semaphore's wait list is held
+    /// by another task, the task releasing permits adds them to this counter.
+    /// The task holding the lock will continue releasing permits until the
+    /// counter is drained, allowing the `release` operation to be wait free.
+    ///
+    /// The first bit of this counter indicates that the semaphore is closing.
+    /// Therefore, all values are shifted over one bit.
+    adding: AtomicUsize,
```

Comment:
How does using an `adding` atomic compare to just acquiring the mutex, especially w/ `parking_lot` enabled. Ideally, we would opt for simpler unless there is a measurable improvement.